src/Microsoft.DotNet.Build.CloudTestTasks/AzureHelper.cs

```diff
@@ -0,0 +1,234 @@
+            {
+                if (headerName.ToLowerInvariant().StartsWith("x-ms-", StringComparison.Ordinal))
+                {
+                    headerNameList.Add(headerName.ToLowerInvariant());
+                }
+            }
+            headerNameList.Sort();
+            foreach (string headerName in headerNameList)
+            {
+                var builder = new StringBuilder(headerName);
+                var separator = ":";
+                foreach (string headerValue in GetHeaderValues(request.Headers, headerName))
+                {
+                    var trimmedValue = headerValue.Replace("\r\n", string.Empty);
+                    builder.Append(separator);
+                    builder.Append(trimmedValue);
+                    separator = ",";
+                }
+                sb.Append(builder);
+                sb.Append("\n");
+            }
+            return sb.ToString();
+        }
+
+        public static ArrayList GetHeaderValues(WebHeaderCollection headers, string headerName)
+        {
+            var list = new ArrayList();
+            var values = headers[headerName];
+            if (values != null)
+            {
+                list.Add(values.TrimStart(null));
+            }
+            return list;
+        }
+
+        public static string GetCanonicalizedResource(Uri address, string accountName)
+        {
+            var str = new StringBuilder();
+            var builder = new StringBuilder("/");
+            builder.Append(accountName);
+            builder.Append(address.AbsolutePath);
+            str.Append(builder);
+            var values2 = new Dictionary<string, HashSet<string>>();
+            if (true)
+            {
+                var values = new Dictionary<string, HashSet<string>>();
+                var newreg = new Regex(@"\?(\w+)\=([\w|\=]+)|\&(\w+)\=([\w|\=]+)");
+                var matches = newreg.Matches(address.Query);
+                foreach (Match match in matches)
+                {
+                    string key, value;
+                    if (match.Groups[1].Value != String.Empty)
+                    {
+                        key = match.Groups[1].Value;
+                        value = match.Groups[2].Value;
+                    }
+                    else
+                    {
+                        key = match.Groups[3].Value;
+                        value = match.Groups[4].Value;
+                    }
+
+                    if (!values.ContainsKey(key))
+                    {
+                        HashSet<string> newSet = new HashSet<string>();
+                        newSet.Add(value);
+                        values.Add(key, newSet);
+                    }
+                    else
+                    {
+                        HashSet<string> setOfValues = null;
+                        values.TryGetValue(key, out setOfValues);
+                        setOfValues.Add(value);
+                        setOfValues.Remove(key);
```

Comment:
We add the value and then remove the key? So if I have "?b=b&b=c", then in the end we only have "c", since I saw the key a time after I saw it's value? I'm like 90% that's unintentional.